### ISSUE TITLE
 Sciprt works on .net but not on mono

### ISSUE THREAD

**Issue by eqr** [fsharp/fsharp#739]
Hello everyone.
I'm trying to do the first example from the [book ](https://books.google.ru/books/about/Machine_Learning_Projects_for_NET_Develo.html?id=VlAnCgAAQBAJ&printsec=frontcover&source=kp_read_button&redir_esc=y#v=onepage&q&f=false). The script I'm trying to run is 
```fsharp
open System.IO
type Observation = { Label:string; Pixels:int[] }
let toObservation (csvData:string) = 
    let columns = csvData.Split(',')
    let label = columns.[0]
    let pixels = columns.[1..] |> Array.map int
    { Label = label; Pixels = pixels }

let reader path = 
    let data = File.ReadAllLines path
    data.[1..]
    |> Array.map toObservation

let manhattanDistance(pixels1:int[], pixels2:int[]) = 
    Array.zip pixels1 pixels2
    |> Array.map (fun (x, y) -> abs(x-y))
    |> Array.sum    

let train (trainingSet:Observation[]) = 
    let classify (pixels:int[]) = 
        trainingSet
        |> Array.minBy (fun x -> manhattanDistance(x.Pixels, pixels))
        |> fun x -> x.Label
    classify

let trainingPath = __SOURCE_DIRECTORY__ + @"../../Digits.Data/trainingsample.csv"
let trainingData = reader trainingPath
let classifier = train trainingData

let validationPath =__SOURCE_DIRECTORY__ + @"../../Digits.Data/validationsample.csv"
let validationData = reader validationPath
validationData
    |> Array.averageBy (fun x -> if classifier x.Pixels = x.Label then 1. else 0.)
    |> printfn "Correct: %.3f"

```
It works well when I run it using fsi.exe on Windows or in Interactive in VS2015. When I try to run it on Linux Mint using fsharpi on mono 5.0 I have the next error: error FS0193: Specified method is not supported.

When I try to run it on MacOs I have the next exception:
```
System.ArrayTypeMismatchException: Attempted to access an element as a type incompatible with the array.
  at (wrapper stelemref) System.Object:virt_stelemref_sealed_class (intptr,object)
  at FSI_0001.reader (System.String path) [0x00174] in <1d85a227534b4146a528f6345878564c>:0
  at <StartupCode$FSI_0001>.$FSI_0001.main@ () [0x00002] in <1d85a227534b4146a528f6345878564c>:0
  at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00032] in <b697cffb61b74023aa9c587e6c49beb3>:0
Stopped due to error
```

I'm new to FSharp world and I'm out of ideas or way to troubleshoot this (paths to files are correct). Ideas are appreciated.

Interesting thing here is that when I compile the script into executable using fsharpc, the executable works fine.

Thanks in advance.

**Comment by dsyme** [comment]
@eqr I'm very surprised this isn't working, the code looks simple enough.  @nosami could you take a look?

**Comment by eqr** [comment]
@dsyme, @nosami 
In case you need data, the files are [here](https://github.com/mathias-brandewinder/machine-learning-projects-for-dot-net-developers/tree/master/chapter-1/DigitsRecognizer/Data)

**Comment by nosami** [comment]
```
13:16 $ ls -la
total 19672
drwxr-xr-x    5 jason  staff      170 May 17 13:16 .
drwxr-xr-x  133 jason  staff     4522 May 17 13:13 ..
-rw-r--r--    1 jason  staff     1098 May 17 13:16 bug739.fsx
-rw-r--r--    1 jason  staff  9145009 May 17 13:15 trainingsample.csv
-rw-r--r--    1 jason  staff   919002 May 17 13:15 validationsample.csv
✔ ~/src/fsharp739
13:16 $ fsharpi bug739.fsx
System.ArrayTypeMismatchException: Attempted to access an element as a type incompatible with the array.
  at (wrapper stelemref) System.Object:virt_stelemref_sealed_class (intptr,object)
  at FSI_0001.reader (System.String path) [0x00174] in <a5ab36e05fd44e2482fe569109a0f869>:0
  at <StartupCode$FSI_0001>.$FSI_0001.main@ () [0x0001a] in <a5ab36e05fd44e2482fe569109a0f869>:0
  at (wrapper managed-to-native) System.Reflection.MonoMethod:InternalInvoke (System.Reflection.MonoMethod,object,object[],System.Exception&)
  at System.Reflection.MonoMethod.Invoke (System.Object obj, System.Reflection.BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) [0x00032] in <b697cffb61b74023aa9c587e6c49beb3>:0
Stopped due to error
```

I get the same error

**Comment by nosami** [comment]
Reflection.Emit log for the above script https://gist.github.com/nosami/832bd0ff2306033d3f0897b00b96af90

**Comment by nosami** [comment]
If I evaluate each expression individually, the script works

**Comment by eqr** [comment]
@nosami Can I assist you somehow in your investigation?

**Comment by nosami** [comment]
> @nosami Can I assist you somehow in your investigation?
I'm not even sure if this is an fsharp/fsharp bug or a mono issue yet :)

**Comment by nosami** [comment]
Working Reflection.Emit log (when ran interactively) https://gist.github.com/nosami/f777919512012d0892c4882f54743038

**Comment by nosami** [comment]
Native stacktrace when running FsiExe.exe from FCS https://gist.github.com/nosami/876a7ca4f2e501263f009acac5aa0006

**Comment by nosami** [comment]
Debug assert is here https://github.com/mono/mono/blob/2017-02/mono/sgen/sgen-scan-object.h#L90

**Comment by nosami** [comment]
This is a smaller repro

```fsharp
type Observation = { Label:string; Pixels:int[] }

let x = [|""|] |> Array.map (fun s -> { Label = s; Pixels =[|1|]})
```

If I run the first line in fsharpi and then the 2nd, it works fine.
If I evaluate both together, it fails

**Comment by nosami** [comment]
Filed https://bugzilla.xamarin.com/show_bug.cgi?id=56611

**Comment by dsyme** [comment]
Thanks!  It looks really important to get this fixed!

**Comment by eqr** [comment]
@nosami Thanks, Jason

**Comment by nosami** [comment]
This is fixed in Mono 5.2.0.164 and later.

**Comment by dsyme** [comment]
@nosami Thanks for the update!
